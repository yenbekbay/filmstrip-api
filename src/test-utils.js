/* @flow */

import _ from 'lodash/fp';
import type from 'type-detect';

const childTypesForArray = _.memoize(_.flow(_.map(type), _.uniq));

const modelFromObject = (obj: ?Object) =>
  _.mapValues(
    _.cond([
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.head, _.eq('Object')),
        ]),
        (val: Array<any>) => [_.flow(_.head, modelFromObject)(val)],
      ],
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.size, _.eq(1)),
        ]),
        (val: Array<any>) =>
          `Array<${_.flow(childTypesForArray, _.head)(val)}>`,
      ],
      [_.flow(type, _.includes(_, ['Object', 'Array'])), modelFromObject],
      [_.stubTrue, type],
    ]),
    obj,
  );

export {modelFromObject}; // eslint-disable-line import/prefer-default-export
