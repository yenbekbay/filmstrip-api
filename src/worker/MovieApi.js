/* @flow */

import { Tmdb, Imdb, Kinopoisk } from 'movie-api';
import _ from 'lodash/fp';

import { tmdbApiKey, imdbUserId } from '../env';
import type { MovieInfo } from '../types';

const TMDB_MOVIE_INFO_QUERY = `
  {
    tmdbId
    imdbId
    backdropUrl
    genres
    keywords
    originalTitle
    originalLanguage
    posterUrl
    productionCountries { iso_3166_1 }
    releaseDate
    runtime
    synopsis
    title
    tmdbRating
    tmdbRatingVoteCount
    credits {
      cast {
        character
        name
        photoUrl
      }
      crew {
        directors {
          name
          photoUrl
        }
      }
    }
    videos {
      iso_639_1
      key
      site
      type
    }
  }
`;
const KP_MOVIE_INFO_QUERY = `
  {
    kpId
    kpRating
    kpRatingVoteCount
    rtCriticsRating
    rtCriticsRatingVoteCount
    mpaaRating
  }
`;

type Query = {
  title: string,
  year: number,
  imdbId: string,
  tmdbId?: ?string,
  kpId?: ?string,
};

class MovieApi {
  _tmdb = new Tmdb({
    apiKey: tmdbApiKey,
    language: 'en',
  });
  _imdb = new Imdb({
    userId: imdbUserId,
  });
  _kp = new Kinopoisk();

  _getTmdbInfo = async ({ imdbId, tmdbId }: Query) => {
    const movieId = tmdbId || await this._tmdb.getMovieId({ imdbId });

    return movieId
      ? this._tmdb.getMovieInfo(movieId, TMDB_MOVIE_INFO_QUERY)
      : {};
  };

  _getKinopoiskInfo = async ({ title, year, kpId }: Query) => {
    const movieId = kpId || await this._kp.getFilmId({ title, year });

    return movieId
      ? this._kp.getFilmInfo(movieId, KP_MOVIE_INFO_QUERY)
      : {};
  };

  getMovieInfo = async (query: Query): Promise<?MovieInfo> => {
    const tmdbInfo = await this._getTmdbInfo(query);

    if (!tmdbInfo) return null;

    const [kpInfo, imdbRating, imdbPopularity] = await Promise.all([
      this._getKinopoiskInfo(query),
      this._imdb.getRating(query.imdbId || tmdbInfo.imdbId),
      this._imdb.getPopularity(query.imdbId || tmdbInfo.imdbId),
    ]);

    return {
      ..._.omit(['videos'], tmdbInfo),
      ...imdbRating,
      ...kpInfo,
      imdbPopularity: (imdbPopularity && imdbPopularity) < 1000
        ? imdbPopularity
        : null,
      productionCountries: _.map('iso_3166_1', tmdbInfo.productionCountries),
      youtubeIds: _.flow(
        _.filter(
          ({ type, site }: Object) => type === 'Trailer' && site === 'YouTube',
        ),
        _.map('key'),
      )(tmdbInfo.videos),
    };
  };

  getUpdates = async (query: Query & { imdbId: string }): Promise<?Object> => {
    const [kpInfo, imdbRating, imdbPopularity] = await Promise.all([
      this._getKinopoiskInfo(query),
      this._imdb.getRating(query.imdbId),
      this._imdb.getPopularity(query.imdbId),
    ]);

    return {
      ...imdbRating,
      ...kpInfo,
      imdbPopularity: (imdbPopularity && imdbPopularity) < 1000
        ? imdbPopularity
        : null,
    };
  };

  findMovie = async (
    query: { title: string, year: string },
  ): Promise<{ tmdbId: number, title: string }> => {
    const res = await this._tmdb._connector.apiGet(
      'search/movie',
      { query: query.title, year: query.year },
    );

    return _.flow(
      _.getOr([], 'results'),
      _.head,
      ({ id, title }: { id: number, title: string } = {}) => ({
        tmdbId: id,
        title,
      }),
    )(res);
  };
}

export default MovieApi;
