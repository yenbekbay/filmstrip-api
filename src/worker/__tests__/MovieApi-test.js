/* @flow */

import { modelFromObject } from '../../test-utils';
import MovieApi from '../MovieApi';

describe('MovieApi', () => {
  let movieApi: MovieApi;

  beforeAll(() => {
    movieApi = new MovieApi();
  });

  it('fetches movie info for a given query', async () => {
    const movieInfo = await movieApi.getMovieInfo({
      title: 'Star Wars: The Force Awakens',
      year: 2015,
      imdbId: 'tt2488496',
    });
    expect(modelFromObject(movieInfo)).toMatchSnapshot();
  }, 10000);
});
