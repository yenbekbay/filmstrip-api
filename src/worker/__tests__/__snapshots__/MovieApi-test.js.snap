exports[`MovieApi fetches movie info for a given query 1`] = `
Object {
  "backdropUrl": "string",
  "credits": Object {
    "en": Object {
      "cast": Array [
        Object {
          "character": "string",
          "name": "string",
          "photoUrl": "string",
        },
      ],
      "crew": Object {
        "cinematographers": Array [
          Object {
            "department": "string",
            "job": "string",
            "name": "string",
            "photoUrl": "null",
          },
        ],
        "composers": Array [
          Object {
            "department": "string",
            "job": "string",
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "directors": Array [
          Object {
            "department": "string",
            "job": "string",
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "producers": Array [
          Object {
            "department": "string",
            "job": "string",
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "writers": Array [
          Object {
            "department": "string",
            "job": "string",
            "name": "string",
            "photoUrl": "string",
          },
        ],
      },
    },
    "ru": Object {
      "cast": Array [
        Object {
          "name": "string",
          "photoUrl": "string",
        },
      ],
      "crew": Object {
        "cinematographers": Array [
          Object {
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "composers": Array [
          Object {
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "directors": Array [
          Object {
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "producers": Array [
          Object {
            "name": "string",
            "photoUrl": "string",
          },
        ],
        "writers": Array [
          Object {
            "name": "string",
            "photoUrl": "string",
          },
        ],
      },
    },
  },
  "genres": Object {
    "en": "Array<string>",
    "ru": "Array<string>",
  },
  "imdbId": "string",
  "imdbPopularity": "number",
  "imdbRating": "number",
  "imdbRatingVoteCount": "number",
  "keywords": "Array<string>",
  "kpId": "number",
  "kpRating": "number",
  "kpRatingVoteCount": "number",
  "mpaaRating": "string",
  "originalLanguage": "string",
  "originalTitle": "string",
  "posterUrl": Object {
    "en": "string",
    "ru": "string",
  },
  "productionCountries": Object {
    "en": "Array<string>",
    "ru": "Array<string>",
  },
  "releaseDate": "string",
  "rtCriticsRating": "number",
  "rtCriticsRatingVoteCount": "number",
  "runtime": "number",
  "synopsis": Object {
    "en": "string",
    "ru": "string",
  },
  "title": Object {
    "en": "string",
    "ru": "string",
  },
  "tmdbId": "number",
  "tmdbRating": "number",
  "tmdbRatingVoteCount": "number",
  "youtubeIds": Object {
    "en": "Array<string>",
    "ru": "Array<string>",
  },
}
`;
